src/itemdef.cpp

```diff
@@ -319,103 +318,72 @@ class CItemDefManager: public IWritableItemDefManager
 	}
 #ifndef SERVER
 public:
-	ClientCached* createClientCachedDirect(const std::string &name,
-			Client *client) const
+	ClientCached* createClientCachedDirect(const ItemStack &item, Client *client) const
 	{
-		infostream<<"Lazily creating item texture and mesh for \""
-				<<name<<"\""<<std::endl;
-
 		// This is not thread-safe
 		sanity_check(std::this_thread::get_id() == m_main_thread);
 
+		const ItemDefinition &def = item.getDefinition(this);
+		std::string inventory_image = item.getInventoryImage(this);
+		std::string inventory_overlay = item.getInventoryOverlay(this);
+		std::string cache_key = def.name;
+		if (!inventory_image.empty())
+			cache_key += "/" + inventory_image;
+		if (!inventory_overlay.empty())
+			cache_key += ":" + inventory_overlay;
+
+		infostream << "Lazily creating item texture and mesh for \""
+				<< cache_key << "\""<<std::endl;
+
 		// Skip if already in cache
-		ClientCached *cc = NULL;
-		m_clientcached.get(name, &cc);
-		if(cc)
-			return cc;
+		auto it = m_clientcached.find(cache_key);
+		if (it != m_clientcached.end())
+			return it->second.get();
 
 		ITextureSource *tsrc = client->getTextureSource();
-		const ItemDefinition &def = get(name);
 
 		// Create new ClientCached
-		cc = new ClientCached();
+		auto cc = std::make_unique<ClientCached>();
 
 		// Create an inventory texture
 		cc->inventory_texture = NULL;
-		if (!def.inventory_image.empty())
-			cc->inventory_texture = tsrc->getTexture(def.inventory_image);
-
-		ItemStack item = ItemStack();
-		item.name = def.name;
-
+		if (!inventory_image.empty())
+			cc->inventory_texture = tsrc->getTexture(inventory_image);
 		getItemMesh(client, item, &(cc->wield_mesh));
```

Comment:
You could here just pass item name, `inventory_image` and `inventory_overlay` (as opposed to `item`), to avoid similar `cache_key` issues in future changes to `getItemMesh`.